coalib/testing/BaseTestHelper.py

```diff
@@ -1,45 +1,43 @@
 
-def _assert_comparable_equal(observed_result, expected_result):
-    """
-    Compares individual fields of the result objects using
-    `__compare_fields__` of `coala_utils.Comparable` class
-    and yields messages in case of an attribute mismatch.
-    """
 
-    if not len(observed_result) == len(expected_result):
-        assert observed_result == expected_result, '%s != %s' % (
-            observed_result, expected_result)
-
-    messages = ''
-    for observed, expected in zip(observed_result, expected_result):
-        if (isinstance(observed, Comparable) and
-            isinstance(expected, Comparable)) and (type(observed) is
-                                                   type(expected)):
-            for attribute in type(observed).__compare_fields__:
-                try:
-                    assert getattr(observed, attribute) == getattr(
-                        expected, attribute), (
-                        '{} mismatch: {}, {} != {}, {}'.format(
-                            attribute,
-                            observed.origin, observed.message,
-                            expected.origin, expected.message))
-                except AssertionError as ex:
-                    messages += (str(ex) + '\n\n')
-        else:
-            assert observed_result == expected_result, '%s != %s' % (
-                observed_result, expected_result)
-
-    if messages:
-        raise AssertionError(messages)
-
-
-class BaseTestHelper:
+class BaseTestHelper(unittest.TestCase):
     """
     This is a base class for all Bears' tests of coala's testing API.
     """
 
+    def _assert_comparable_equal(self,
+                                 observed_result,
+                                 expected_result):
+        """
+        Compares individual fields of the result objects using
+        `__compare_fields__` of `coala_utils.Comparable` class
+        and raises messages in case of an attribute mismatch.
+        """
+
+        if not len(observed_result) == len(expected_result):
+            self.assertEqual(observed_result, expected_result)
+
+        messages = []
+        for observed, expected in zip(observed_result, expected_result):
+            if (isinstance(observed, Comparable) and
+                isinstance(expected, Comparable)) and (
+                    Comparable.__eq__(expected, observed)):
+                for attribute in observed.__compare_fields__:
+                    try:
+                        self.assertEqual(getattr(observed, attribute),
+                                         getattr(expected, attribute),
+                                         msg='{} mismatch.'.format(attribute))
+                    except AssertionError as ex:
+                        messages.append(str(ex))
+            else:
+                self.assertEqual(observed, expected)
+
+        if messages:
+            raise AssertionError('\n\n'.join(message for message in messages))
```

Comment:
When running `test_jinja2_templating` (due to https://github.com/coala/coala-bears/issues/2678) which isnt the best of test approaches, but far from the worst, here are my error results

```pytest
>               raise AssertionError(messages)
E               AssertionError: 'Usin[68 chars]o XSS. Ensure autoescape=True to mitigate XSS vulnerabilities.' != 'Usin[68 chars]o XSS. Use autoescape=True to mitigate XSS vulnerabilities.'
E               - Using jinja2 templates with autoescape=False is dangerous and can lead to XSS. Ensure autoescape=True to mitigate XSS vulnerabilities.
E               ?                                                                                ^^ --
E               + Using jinja2 templates with autoescape=False is dangerous and can lead to XSS. Use autoescape=True to mitigate XSS vulnerabilities.
E               ?                                                                                ^
E                : message_base mismatch.

../coala/coalib/testing/LocalBearTestHelper.py:136: AssertionError
```

Now how can I work out which of the five `Result` that was about?

This entire test system is fundamentally flawed.

test helpers should help diagose, not help write tests which are unable to be diagnosed.

If I change it to `assert sorted(bear_output) == sorted(results)`, I see *nothing* because these `Comparable` objects are impossible for pytest to automatically inspect dynamically.  And `self.assertCountEqual` has also been busted: the output is a mess.  We are breaking the standard Python testing tools, and replacing them with something less helpful.

So at the moment, with pytest and unittest builtin systems not working on `Result` objects, we are reliant on this `Comparable` class and this helper to actually help decode the problem.  It doesnt do that before this patch, and this patch doesnt make it much better.